Default to no fsc

som-snytt opened the pull request:
Let the script runner default to not firing up a compilation daemon.

Accept `-howtorun:fsc` to mean use the server.

Take `-nc` to mean shut it down. This is much more convenient than
using `fsc -shutdown`.

Edit: Use `-Yscriptrunner flavor`, where flavor is `default`, `resident`, `shutdown`, or a class name.

For `shutdown` to run, use `scala -Yscriptrunner shutdown -e ""`. Without `-e` or a file, it goes to REPL. An empty filename `""` is not sufficient.

som-snytt committed:
Default to no fsc

Let the script runner default to not firing up a compilation daemon.

Accept `-howtorun:fsc` to mean use the server.

Take `-nc` to mean shut it down. This is much more convenient than
using `fsc -shutdown`.

SethTisue commented:
are you spying on Scala team meetings? 👀 we just discussed/suggested this this morning.

in the longer run it might be nice to just rip up all the compilation-daemon stuff and throw it away, but we're running out of time for would-be-nices for 2.13.

SethTisue commented:
👍 defaulting to `-nc`, but I'm less certain about the change to be more aggressive with the shutting-down. not sure it makes sense to rock the boat there, maybe better to leave well enough alone? not sure.

lrytz commented on src/compiler/scala/tools/nsc/GenericRunnerSettings.scala:
```diff
@@ -16,8 +16,9 @@ class GenericRunnerSettings(error: String => Unit) extends Settings(error) {
       "how",
       "how to run the specified code",
-      List("object", "script", "jar", "repl", "guess"),
+      List("object", "script", "jar", "repl", "fsc", "guess"),
```
I think `-howtorun:fsc` is not easy to understand. Maybe `script-compdaemon` is more clear?

som-snytt replied:
I considered `-howtorun:resident` because `fsc` is a brand name. When `fsc` is a module, it will need the brand exposure, but `resident` is generic; ideally, there will be multiple REPL frontends and backends, and some of the backends will support `resident` flag. Or am I getting ahead of myself?

lrytz replied:
I like `script` in the name because that's "how it runs", as a script, just in a special way. alternatively add a `-compdaemon` setting, but i think we want to get rid of it entirely, so i would not add something.

I also couldn't find out how to negate the negation, i.e. `-nc:false` or so, is that possible?

som-snytt replied:
I think `-nc:false` just works.

lrytz replied:
Ah... Here's what tripped me up:

```
➜  sandbox git:(2.13.x) scala -nc Test.scala
hi
➜  sandbox git:(2.13.x) fsc -shutdown
[No compilation server running.]
➜  sandbox git:(2.13.x) scala Test.scala -nc
hi
➜  sandbox git:(2.13.x) fsc -shutdown
[Compile server exited]
➜  sandbox git:(2.13.x) fsc -shutdown
[No compilation server running.]
➜  sandbox git:(2.13.x)
```

som-snytt replied:
I'm doing a quick refactor of ScriptRunner to break out the functionality, and then `-Yscriptrunner:com.github.somsnytt.BestScriptRunnerEver`.

som-snytt replied:
I remember there's a comment in settings.process about handling trailing options; I'll look if it's broken for Generic...Settings.

lrytz commented on src/compiler/scala/tools/nsc/GenericRunnerSettings.scala:
```diff
@@ -45,9 +46,10 @@ class GenericRunnerSettings(error: String => Unit) extends Settings(error) {
       "-nc",
-      "do not use the fsc compilation daemon") withAbbreviation "-nocompdaemon" withPostSetHook((x: BooleanSetting) => {_useCompDaemon = !x.value })
+      "do not use the fsc compilation daemon and shut it down if it is running").withAbbreviation("-nocompdaemon")
+      //.withDeprecationMessage("scripts use cold compilation by default; use -howtorun:fsc to start the compilation daemon")
```
I'm in favor of adding the deprecation warning, so we can remove `-nc` eventually

som-snytt replied:
I'll need a better story for replacing `fsc -shutdown`. Besides `-howtorun:shutdown`. Not really sure anyone ever, anywhere, has ever used `-howtorun`; I think I tweaked it once. Maybe it should be renamed to `-whattorun`.

som-snytt committed:
Refactor ScriptRunner

ScriptRunner is an interface, with standard
implementations for immediate compilation and
resident compile server. A daemon killer
knows how to shutdown the compile server.

som-snytt committed:
Yscriptrunner specifies how to run script

Accept default, resident, shutdown, or custom class
with a constructor that takes settings.

`-nc` is a deprecated alias.

lrytz commented on src/repl-frontend/scala/tools/nsc/MainGenericRunner.scala:
```diff
@@ -62,9 +62,9 @@ class MainGenericRunner {
           ObjectRunner.runAndCatch(settings.classpathURLs, thingToRun, command.arguments)
         case AsScript if isE =>
-          Right(ScriptRunner.runCommand(settings, combinedCode, thingToRun +: command.arguments))
+          Right(!ScriptRunner(settings).runScriptText(combinedCode, thingToRun +: command.arguments).isDefined)
```
why do we discard the exception here?

som-snytt replied:
You never let me get away with anything. I was worn down by the refactor at that point. And by figuring out what to call it besides `runCommand`. I'll take another look at how the values bubble up; there were other spots like `CommonRunner` IIRC.

adriaanm replied:
Perhaps along with the rebase?

lrytz reviewed (approved):
LGTM, thanks!

som-snytt committed:
Use Option[Throwable] for Either[Thr,Boolean]

We don't seem to care if something ran unsuccessfully
without throwing.

Also, leverage NonFatal when throwing yourself
on a hand grenade.

som-snytt commented:
It could use another round. In ScriptRunner, withCompiledScript and runCompiled return boolean instead of passing back the exception; the handler body is also boolean-valued. Those could be flipped to `Option[Bomb]`, but rebasing incurs a merge from my previous helpful efforts.

som-snytt commented on src/compiler/scala/tools/nsc/util/Exceptional.scala:
```diff
@@ -15,8 +14,4 @@ object Exceptional {
-  // partest still uses the old name.  once we re-in-source partest we
-  // can get rid of this, but in the meantime, we'd rather not branch
-  // partest just because one method got renamed
-  def unwrap(x: Throwable): Throwable = rootCause(x)
```
Merit badge territory! :medal_military:

adriaanm replied:
:100:

som-snytt commented on src/repl-frontend/scala/tools/nsc/MainGenericRunner.scala:
```diff
@@ -58,17 +58,17 @@ class MainGenericRunner {
           JarRunner.runJar(settings, thingToRun, command.arguments)
         case Error =>
-          Right(false)
+          None
```
This changes the exit value of `scala Junk` from `1` to `0`. `howToRun` sees `no such class`.

Are there conditions where the other runners did not throw but would return `Some(false)` for failure? Check failed compilation, for example.
